### PR TITLE
Add torchmultimodal tutorial for flava finetuning

### PR DESCRIPTION
**Pull request by ankitade**
Adding first tutorial for torchmultimodal around how to finetune FLAVA for vqa.

### CODE REVIEW

File: beginner_source/flava_finetuning_tutorial.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+# In this tutorial, we will demonstrate how to use a **pretrained SoTA
+# model called** `FLAVA <https://arxiv.org/pdf/2112.04482.pdf>`__ **from
+# TorchMultimodal library to finetune on a multimodal task i.e. visual
+# question answering** (VQA).
+# 
+
+
+######################################################################
+# Installations
+#
```
Comment (svekars) [inline beginner_source/flava_finetuning_tutorial.py:24]:
```suggestion
# -----------------
#
```

File: beginner_source/flava_finetuning_tutorial.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+# TorchMultimodal library to finetune on a multimodal task i.e. visual
+# question answering** (VQA).
+# 
+
+
+######################################################################
+# Installations
+# 
+# We will use TextVQA dataset from HuggingFace for this
+# tutorial. So we install datasets in addition to TorchMultimodal
```
Comment (svekars) [inline beginner_source/flava_finetuning_tutorial.py:26]:
```suggestion
# tutorial. We install datasets in addition to TorchMultimodal.
```

File: beginner_source/flava_finetuning_tutorial.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+# For this tutorial, we treat VQA as a classification task. So we need to
+# download the vocab file with answer classes and create the answer to
+# label mapping.
+# 
+# We also load the `textvqa
+# dataset <https://arxiv.org/pdf/1904.08920.pdf>`__ from HuggingFace
+# 
+
+!wget http://dl.fbaipublicfiles.com/pythia/data/vocab.tar.gz
+!tar xf vocab.tar.gz
```
Comment (svekars) [inline beginner_source/flava_finetuning_tutorial.py:50]: You added this to the Makefile above I believe

File: beginner_source/flava_finetuning_tutorial.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+######################################################################
+# Installations
+# 
+# We will use TextVQA dataset from HuggingFace for this
+# tutorial. So we install datasets in addition to TorchMultimodal
+# 
+
+# TODO: replace with install from pip when binary is ready
+!git clone https://github.com/facebookresearch/multimodal.git
+!pip install -r multimodal/requirements.txt
```
Comment (svekars) [inline beginner_source/flava_finetuning_tutorial.py:31]: Should go to requirements.txt - I see only two items in that file - you can add them to the requirements.txt

File: beginner_source/flava_finetuning_tutorial.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+import os
+import sys 
+sys.path.append(os.path.join(os.getcwd(),"multimodal"))
+sys.path.append(os.getcwd())
+!pip install datasets
+!pip install transformers
```
Comment (svekars) [inline beginner_source/flava_finetuning_tutorial.py:37]: already in requirements.txt

File: beginner_source/flava_finetuning_tutorial.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+# 
+# We also load the `textvqa
+# dataset <https://arxiv.org/pdf/1904.08920.pdf>`__ from HuggingFace
+# 
+
+!wget http://dl.fbaipublicfiles.com/pythia/data/vocab.tar.gz
+!tar xf vocab.tar.gz
+
+
+with open("vocabs/answers_textvqa_more_than_1.txt") as f:
```
Comment (svekars) [inline beginner_source/flava_finetuning_tutorial.py:53]: this should go to where you have downloaded your data - probably 'data/'

File: requirements.txt [hunk 7]
```diff
@@ -45,3 +48,7 @@ wget
 gym==0.24.0
 gym-super-mario-bros==7.3.0
 timm
+
+# flava tutorial - multimodal
+packaging
```
Comment (langong347) [inline requirements.txt:53]: I don't think we need `packaging` anymore. I removed it in this [PR](https://github.com/facebookresearch/multimodal/pull/322/files#diff-4d7c51b1efe9043e44439a949dfd92e5827321b34082903477fd04876edb7552).

File: beginner_source/flava_finetuning_tutorial.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+# Conclusion
+#
+# This tutorial introduced the basics around how to finetune on a
+# multimodal task using FLAVA from TorchMultimodal. Please also check out
+# other examples from the library like
+# `MDETR <https://github.com/facebookresearch/multimodal/tree/main/torchmultimodal/models/mdetr>`__
+# which is a multimodal model for object detection and
+# `Omnivore <https://github.com/facebookresearch/multimodal/blob/main/torchmultimodal/models/omnivore.py>`__
+# which is multitask model spanning image, video and 3d classification.
+#
```
Comment (langong347) [inline beginner_source/flava_finetuning_tutorial.py:162]: Did we miss the generation example links?

https://github.com/facebookresearch/multimodal/blob/main/examples/mugen/generation/text_video_gpt.py
https://github.com/facebookresearch/multimodal/blob/main/torchmultimodal/models/gpt.py

File: beginner_source/flava_finetuning_tutorial.py [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+print(vocab[:5])
+
+from datasets import load_dataset
+dataset = load_dataset("textvqa")
+
+from IPython.display import display, Image
+idx = 5
+print("Question: ", dataset["train"][idx]["question"])
+print("Answers: " ,dataset["train"][idx]["answers"])
+display(dataset["train"][idx]["image"].resize((500,500)))
```
Comment (svekars) [inline beginner_source/flava_finetuning_tutorial.py:68]: Need to find a way to render this image correctly

File: beginner_source/flava_finetuning_tutorial.py [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+######################################################################
+# For this tutorial, we treat VQA as a classification task. So we need to
+# download the vocab file with answer classes and create the answer to
+# label mapping.
+#
+# We also load the `textvqa
+# dataset <https://arxiv.org/pdf/1904.08920.pdf>`__ from HuggingFace
+#
+
+with open("data/vocabs/answers_textvqa_more_than_1.txt") as f:
```
Comment (suraj813) [inline beginner_source/flava_finetuning_tutorial.py:45]: Not sure where this file is from. Does the user need to download this or can it be generated from the dataset?

File: beginner_source/flava_finetuning_tutorial.py [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+!pip install transformers
+"""
+
+######################################################################
+# For this tutorial, we treat VQA as a classification task. So we need to
+# download the vocab file with answer classes and create the answer to
+# label mapping.
+#
+# We also load the `textvqa
+# dataset <https://arxiv.org/pdf/1904.08920.pdf>`__ from HuggingFace
```
Comment (suraj813) [inline beginner_source/flava_finetuning_tutorial.py:42]: Add 1 sentence describing this dataset
```suggestion
# dataset <https://arxiv.org/pdf/1904.08920.pdf>`__  containing 45,336 questions on 28,408 textual images from HuggingFace.
```

File: beginner_source/flava_finetuning_tutorial.py [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+######################################################################
+# Multimodal AI has recently become very popular owing to its ubiquitous
+# nature, from use cases like image captioning and visual search to more
+# recent applications like image generation from text. **TorchMultimodal
+# is a library powered by Pytorch consisting of building blocks and end to
+# end examples, aiming to enable and accelerate research in
+# multimodality**.
+# 
+# In this tutorial, we will demonstrate how to use a **pretrained SoTA
+# model called** `FLAVA <https://arxiv.org/pdf/2112.04482.pdf>`__ **from
```
Comment (suraj813) [inline beginner_source/flava_finetuning_tutorial.py:16]: An overview of the FLAVA model would be helpful here (maybe with Fig. 2 in the paper?)

File: beginner_source/flava_finetuning_tutorial.py [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+# by uncommenting the following:
+#
+"""
+!pip install torchmultimodal-nightly
+!pip install datasets
+!pip install transformers
+"""
+
+######################################################################
+# For this tutorial, we treat VQA as a classification task. So we need to
```
Comment (suraj813) [inline beginner_source/flava_finetuning_tutorial.py:37]: Needs some more detail on the classification task, maybe something like
```suggestion
# For this tutorial, we treat VQA as a classification task where the inputs are images containing text and the outputs are single-word answers.
```

File: beginner_source/flava_finetuning_tutorial.py (hunk 9 above)
Comment (svekars) [reply beginner_source/flava_finetuning_tutorial.py:68]: @malfet any suggestions on how to make the image render correctly. Right now this is what I see if I run it locally:

Screen Shot 2022-09-27 at 2 55 07 PM (2)

File: beginner_source/flava_finetuning_tutorial.py [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+
+train_dataloader = DataLoader(dataset["train"], batch_size= BATCH_SIZE)
+optimizer = torch.optim.AdamW(model.parameters())
+
+
+epochs = 1
+for _ in range(epochs):
+  for idx, batch in enumerate(train_dataloader):
+    optimizer.zero_grad()
+    out = model(text = batch["input_ids"], image = batch["image"], labels = batch["answers"], required_embedding="mm")
```
Comment (suraj813) [inline beginner_source/flava_finetuning_tutorial.py:144]: What is the `required_embedding` arg doing? It is not as obvious as the other params, maybe add a note in the plaintext above

Comment (suraj813) [inline beginner_source/flava_finetuning_tutorial.py:144]: Does this need retraining the encoders too, or just the head?

**Review by suraj813 — COMMENTED** [review]
Left some feedback for more detail, and some suggested changes

File: beginner_source/flava_finetuning_tutorial.py (hunk 9 above)
Comment (svekars) [reply beginner_source/flava_finetuning_tutorial.py:68]: This worked but someone might have a better solution:
```
from IPython.display import display, Image
import matplotlib.pyplot as plt
import numpy as np

idx = 5
print("Question: ", dataset["train"][idx]["question"])
print("Answers: " ,dataset["train"][idx]["answers"])
im = np.asarray(dataset["train"][idx]["image"].resize((500,500)))
plt.imshow(im)
plt.show()
```
Screen Shot 2022-09-29 at 9 50 45 AM (2)

File: beginner_source/flava_finetuning_tutorial.py (hunk 5 above)
Comment (malfet) [reply beginner_source/flava_finetuning_tutorial.py:33]: Yes, but we want it to be present in the notebook

File: beginner_source/flava_finetuning_tutorial.py [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+
+######################################################################
+# For this tutorial, we treat VQA as a classification task. So we need to
+# download the vocab file with answer classes and create the answer to
+# label mapping.
+#
+# We also load the `textvqa
+# dataset <https://arxiv.org/pdf/1904.08920.pdf>`__ from HuggingFace
+#
+
```
Comment (svekars) [inline beginner_source/flava_finetuning_tutorial.py:44]: Let's add this here:
```
# Steps
# -----
#
# 1. Download the HuggingFace dataset to a directory on your computer by
#    running the following command:
#
#    .. code-block::
#
#       wget http://dl.fbaipublicfiles.com/pythia/data/vocab.tar.gz
#       tar xf vocab.tar.gz
#
#    If you are running this tutorial in Google Colab, run these commands
#    in a new cell and prepend these commands with an exclamation mark (!).
#
# 2. Next, let's load the dataset and see how many answer classes there
#    are in the dataset.
```

File: beginner_source/flava_finetuning_tutorial.py (hunk 10 above)
Comment (malfet) [reply beginner_source/flava_finetuning_tutorial.py:45]: It's downloaded, see `Makefile` changes

File: beginner_source/flava_finetuning_tutorial.py [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+for idx, entry in enumerate(vocab):
+  answer_to_idx[entry.strip("\n")] = idx
+
+
+######################################################################
+# We see there are 3997 answer classes including a class representing
+# unknown answers
+#
+
+print(len(vocab))
```
Comment (svekars) [inline beginner_source/flava_finetuning_tutorial.py:58]: Suggestion: move line 58-59 to under line 50 so it's one block.

File: beginner_source/flava_finetuning_tutorial.py [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+  vocab = f.readlines()
+
+answer_to_idx = {}
+for idx, entry in enumerate(vocab):
+  answer_to_idx[entry.strip("\n")] = idx
+
+
+######################################################################
+# We see there are 3997 answer classes including a class representing
+# unknown answers
```
Comment (svekars) [inline beginner_source/flava_finetuning_tutorial.py:55]:
```suggestion
# unknown answers.
```

File: beginner_source/flava_finetuning_tutorial.py [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+#
+
+print(len(vocab))
+print(vocab[:5])
+
```
Comment (svekars) [inline beginner_source/flava_finetuning_tutorial.py:60]: This has to be step 3: 
```
#################################################
# 3. Next, let's blah (add why are we doing what's on line 61 -68)

File: beginner_source/flava_finetuning_tutorial.py [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+from IPython.display import display, Image
+idx = 5
+print("Question: ", dataset["train"][idx]["question"])
+print("Answers: " ,dataset["train"][idx]["answers"])
+display(dataset["train"][idx]["image"].resize((500,500)))
+
+
+######################################################################
+# Next we write the transform function to convert the image and text into
```
Comment (svekars) [inline beginner_source/flava_finetuning_tutorial.py:72]:
```suggestion
# 4. Next, we write the transform function to convert the image and text into
```

File: beginner_source/flava_finetuning_tutorial.py [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+  batch["answers"] = torch.as_tensor([ans_idx])
+  return batch
+
+tokenizer=BertTokenizer.from_pretrained("bert-base-uncased",padding="max_length",max_length=512)
+transform=partial(transform,tokenizer)
+dataset.set_transform(transform)
+
+
+######################################################################
+# Finally, we import the flava_model_for_classification from
```
Comment (svekars) [inline beginner_source/flava_finetuning_tutorial.py:110]:
```suggestion
# 5. Finally, we import the flava_model_for_classification from
```

File: beginner_source/flava_finetuning_tutorial.py [hunk 21]
```diff
@@ -0,0 +1,163 @@
...
+    loss.backward()
+    optimizer.step()
+    print(f"Loss at step {idx} = {loss}")
+    if idx > MAX_STEPS-1:
+      break
+
+
+######################################################################
+# Conclusion
+#
```
Comment (svekars) [inline beginner_source/flava_finetuning_tutorial.py:155]:
```suggestion
# -------------------
#
```

File: beginner_source/flava_finetuning_tutorial.py [hunk 22]
```diff
@@ -0,0 +1,163 @@
...
+# 
+# In this tutorial, we will demonstrate how to use a **pretrained SoTA
+# model called** `FLAVA <https://arxiv.org/pdf/2112.04482.pdf>`__ **from
+# TorchMultimodal library to finetune on a multimodal task i.e. visual
+# question answering** (VQA).
+# 
+
+
+######################################################################
+# Installations
```
Comment (svekars) [inline beginner_source/flava_finetuning_tutorial.py:23]:
```suggestion
# Installation
```

File: beginner_source/flava_finetuning_tutorial.py [hunk 23]
```diff
@@ -0,0 +1,163 @@
...
+#
```
Comment (svekars) [inline beginner_source/flava_finetuning_tutorial.py:24]:
```suggestion
# -----------------
# 
```

File: beginner_source/flava_finetuning_tutorial.py (hunk 9 above)
Comment (malfet) [reply beginner_source/flava_finetuning_tutorial.py:68]: I'm a bit unsure, how you've rendered it? But traditionally we use matplotlib rather than IPython

File: beginner_source/flava_finetuning_tutorial.py (hunk 10 above)
Comment (suraj813) [reply beginner_source/flava_finetuning_tutorial.py:45]: The Makefile download  works for building the tutorial, but not for a user running this tutorial on a colab/local notebook? Should the `tutorial.py` have a comment with the vocab url?

**Comment by svekars** [comment]
@ankitade what's the status on this? Can you resolve the merge conflict?

File: beginner_source/flava_finetuning_tutorial.py (hunk 8 above)
Comment (ankitade) [reply beginner_source/flava_finetuning_tutorial.py:190]: i can add it in follow up PR

**Review by langong347 — APPROVED** [review]

File: beginner_source/flava_finetuning_tutorial.py [hunk 24]
```diff
@@ -0,0 +1,190 @@
...
+
+######################################################################
+# Multimodal AI has recently become very popular owing to its ubiquitous
+# nature, from use cases like image captioning and visual search to more
+# recent applications like image generation from text. **TorchMultimodal
+# is a library powered by Pytorch consisting of building blocks and end to
+# end examples, aiming to enable and accelerate research in
+# multimodality**.
+# 
+# In this tutorial, we will demonstrate how to use a **pretrained SoTA
```
Comment (ebsmothers) [inline beginner_source/flava_finetuning_tutorial.py:15]: nit: can we just say state-of-the-art here?

File: beginner_source/flava_finetuning_tutorial.py [hunk 25]
```diff
@@ -0,0 +1,190 @@
...
+# model called** `FLAVA <https://arxiv.org/pdf/2112.04482.pdf>`__ **from
+# TorchMultimodal library to finetune on a multimodal task i.e. visual
+# question answering** (VQA). The model consists of two unimodal transformer
+# based encoders for text and image and a multimodal encoder to combine
+# the two embeddings. It is pretrained using contrastive, image text matching and
```
Comment (ebsmothers) [inline beginner_source/flava_finetuning_tutorial.py:20]: Can the losses be enumerated in a different way here? I feel the comma placement makes this kinda confusing

**Review by ebsmothers — APPROVED** [review]
LGTM, just a couple nits

File: requirements.txt [hunk 26]
```diff
@@ -27,6 +26,9 @@ pytorch-lightning
 torchx
 ax-platform
 nbformat>=4.2.0
+datasets
+transformers
+torchmultimodal-nightly
```
Comment (svekars) [inline requirements.txt:31]: can this be updated to use stable?
